File: app/src/main/java/statusbar/finder/LrcGetter.java
package statusbar.finder;

import android.content.Context;
import android.media.MediaMetadata;
import androidx.core.util.Pair;
import cn.zhaiyifan.lyric.LyricUtils;
import cn.zhaiyifan.lyric.model.Lyric;
import com.github.houbb.opencc4j.util.ZhConverterUtil;
import com.moji4j.MojiConverter;
import com.moji4j.MojiDetector;
import statusbar.finder.data.DataOrigin;
import statusbar.finder.data.LyricResult;
import statusbar.finder.data.MediaInfo;
import statusbar.finder.livedata.LyricsResultChange;
import statusbar.finder.misc.CheckLanguageUtil;
import statusbar.finder.provider.ILrcProvider;
import statusbar.finder.provider.KugouProvider;
import statusbar.finder.provider.MusixMatchProvider;
import statusbar.finder.provider.NeteaseProvider;
import statusbar.finder.provider.QQMusicProvider;
import statusbar.finder.provider.utils.LyricSearchUtil;
import statusbar.finder.sql.ActiveManager;
import statusbar.finder.sql.OriginManager;
import statusbar.finder.sql.ResManager;

import java.io.IOException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Objects;

public class LrcGetter {
    private static final ILrcProvider[] providers = {
            new NeteaseProvider(),
            new KugouProvider(),
            new QQMusicProvider(),
            new MusixMatchProvider()
    };
    private static MessageDigest messageDigest;

    public static Lyric getLyric(Context context, MediaMetadata mediaMetadata, String sysLang, String packageName) {
        return getLyric(context, new MediaInfo(mediaMetadata), sysLang, packageName);
    }

    public static Lyric getLyric(Context context, MediaInfo mediaInfo, String sysLang, String packageName) {
        DatabaseHelper.INSTANCE.init(context);
        MojiDetector detector = new MojiDetector();
        MojiConverter converter = new MojiConverter();

        if (messageDigest == null) {
            try {
                messageDigest = MessageDigest.getInstance("SHA");
            } catch (NoSuchAlgorithmException e) {
                e.fillInStackTrace();
                return null;
            }
        }
        Pair<LyricResult, Long> databaseResult = getActiveLyricFromDatabase(mediaInfo, packageName);
        LyricResult currentResult = databaseResult.first;

        if (currentResult == null) {
            searchLyricsResultByInfo(mediaInfo, databaseResult.second, sysLang);
            currentResult = getActiveLyricFromDatabaseByOriginId(databaseResult.second);
        } else {
            LyricsResultChange.Companion.getInstance().notifyResult(new LyricsResultChange.Data(mediaInfo, currentResult));
            return LyricUtils.parseLyric(currentResult);
        }

        if (currentResult == null && (!detector.hasKana(mediaInfo.getTitle()) && detector.hasLatin(mediaInfo.getTitle()))) {
            MediaInfo hiraganaMediaInfo = mediaInfo.clone();
            hiraganaMediaInfo.setTitle(converter.convertRomajiToHiragana(mediaInfo.getTitle()));
            if (detector.hasLatin(hiraganaMediaInfo.getTitle())) {
                LyricsResultChange.Companion.getInstance().notifyResult(new LyricsResultChange.Data(mediaInfo, null));
                return null;
            }
            searchLyricsResultByInfo(hiraganaMediaInfo, databaseResult.second, sysLang);
            currentResult = getActiveLyricFromDatabaseByOriginId(databaseResult.second);
            if (currentResult == null) {
                hiraganaMediaInfo.setTitle(converter.convertRomajiToKatakana(mediaInfo.getTitle()));
                searchLyricsResultByInfo(hiraganaMediaInfo, databaseResult.second, sysLang);
                currentResult = getActiveLyricFromDatabaseByOriginId(databaseResult.second);
            }
        }


        if (currentResult != null) {
            currentResult.setDataOrigin(DataOrigin.INTERNET);
            LyricsResultChange.Companion.getInstance().notifyResult(new LyricsResultChange.Data(mediaInfo, currentResult));
            return LyricUtils.parseLyric(currentResult);
        } else {
            LyricsResultChange.Companion.getInstance().notifyResult(new LyricsResultChange.Data(mediaInfo, null));
            return null;
        }
    }

    private static void searchLyricsResultByInfo(MediaInfo mediaInfo, Long originId, String sysLang) {
        String bestMatchSource = null;
        long bestMatchDistance = 0;
        for (ILrcProvider provider : providers) {
            try {
                LyricResult lyricResult = provider.getLyric(mediaInfo);
                if (lyricResult != null && lyricResult.getLyric() != null) {
                    String allLyrics;
                    if (lyricResult.getTranslatedLyric() != null) {
                        allLyrics = LyricUtils.getAllLyrics(false, lyricResult.getTranslatedLyric());
                    } else {
                        allLyrics = LyricUtils.getAllLyrics(false, lyricResult.getLyric());
                    }
                    if (!CheckLanguageUtil.isJapanese(allLyrics)) {
                        switch (sysLang) {
                            case "zh-CN":
                                if (lyricResult.getTranslatedLyric() != null) {
                                    lyricResult.setTranslatedLyric(ZhConverterUtil.toSimple(lyricResult.getTranslatedLyric()));
                                } else {
                                    lyricResult.setLyric(ZhConverterUtil.toSimple(lyricResult.getLyric()));
                                }
                                break;
                            case "zh-TW":
                                if (lyricResult.getTranslatedLyric() != null) {
                                    lyricResult.setTranslatedLyric(ZhConverterUtil.toTraditional(lyricResult.getTranslatedLyric()));
                                } else {
                                    lyricResult.setLyric(ZhConverterUtil.toTraditional(lyricResult.getLyric()));
                                }
                            default:
                                break;
                        }
                    }
                    ResManager.INSTANCE.insertResData(originId, lyricResult);
                    if (LyricSearchUtil.isLyricContent(lyricResult.getLyric())
                            && (bestMatchSource == null || bestMatchDistance > lyricResult.getDistance())) {
                        bestMatchSource = lyricResult.getSource();
                    }
                }
            } catch (IOException e) {
                e.fillInStackTrace();
            }
        }
        if (bestMatchSource != null) {
            Res bestMatchResult = ResManager.INSTANCE.getResByOriginIdAndProvider(originId, bestMatchSource);
            assert bestMatchResult != null;
            ActiveManager.INSTANCE.insertActiveLog(originId, bestMatchResult.getId());
        }
    }

    private static Pair<LyricResult, Long> getActiveLyricFromDatabase(MediaInfo mediaInfo, String packageName) {
        Long originId = OriginManager.insertOrGetMediaInfoId(mediaInfo, packageName);
        LyricResult lyricResult = getActiveLyricFromDatabaseByOriginId(originId);
        return new Pair<>(lyricResult, originId);
    }

    private static LyricResult getActiveLyricFromDatabaseByOriginId(Long originId) {
        Long resultId = ActiveManager.INSTANCE.getResultIdByOriginId(originId);
        if (resultId == null) {return null;}
        Res res = ResManager.INSTANCE.getResById(resultId);
        if (res == null) {return null;}
        return new LyricResult(res);
    }
}

File: app/src/main/java/statusbar/finder/LrcGetter.kt
package statusbar.finder

import android.content.Context
import android.media.MediaMetadata
import cn.zhaiyifan.lyric.LyricUtils
import cn.zhaiyifan.lyric.model.Lyric
import com.github.houbb.opencc4j.util.ZhConverterUtil
import statusbar.finder.app.event.LyricsResultChange
import statusbar.finder.app.event.LyricsResultChange.Companion.getInstance
import statusbar.finder.data.db.DatabaseHelper.init
import statusbar.finder.data.model.DataOrigin
import statusbar.finder.data.model.LyricResult
import statusbar.finder.data.model.MediaInfo
import statusbar.finder.data.repository.ActiveRepository.insertActiveLog
import statusbar.finder.data.repository.LyricRepository.getActiveLyricFromDatabase
import statusbar.finder.data.repository.LyricRepository.getActiveLyricFromDatabaseByOriginId
import statusbar.finder.data.repository.ResRepository.getResByOriginIdAndProvider
import statusbar.finder.data.repository.ResRepository.insertResData
import statusbar.finder.modifier.*
import statusbar.finder.provider.KugouProvider
import statusbar.finder.provider.MusixMatchProvider
import statusbar.finder.provider.NeteaseProvider
import statusbar.finder.provider.QQMusicProvider
import statusbar.finder.utils.CheckLanguageUtil
import statusbar.finder.utils.LyricSearchUtil
import java.io.IOException

/**
 * LyricGetterExt - statusbar.finder
 * @description TODO: coming soon.
 * @author VictorModi
 * @email victormodi@outlook.com
 * @date 2025/3/13 17:23
 */
object LrcGetter {
    private val providers = arrayOf(
        NeteaseProvider(),
        KugouProvider(),
        QQMusicProvider(),
        MusixMatchProvider()
    )

    private val modifiers = arrayOf(
        AliasModifier(),
        SimplifiedModifier(),
        RemoveParenthesesModifier(),
        OriginalModifier(),
        KatakanaModifier(),
        HiraganaModifier(),
    )

    fun getLyric(context: Context, mediaMetadata: MediaMetadata, sysLang: String, packageName: String): Lyric? {
        return getLyric(context, MediaInfo(mediaMetadata), sysLang, packageName)
    }

    private fun getLyric(context: Context, mediaInfo: MediaInfo, sysLang: String, packageName: String): Lyric? {
        init(context)
        val databaseResult = getActiveLyricFromDatabase(mediaInfo, packageName)
        var currentResult: LyricResult? = databaseResult.first
        currentResult?.let {
            getInstance().notifyResult(LyricsResultChange.Data(mediaInfo, it))
            return LyricUtils.parseLyric(it, mediaInfo, packageName)
        }
        for (modifier in modifiers) {
            modifier.modify(mediaInfo, databaseResult.second)?.let {
                searchLyricsResultByInfo(
                    it,
                    databaseResult.second,
                    sysLang,
                )
            }
            currentResult = getActiveLyricFromDatabaseByOriginId(databaseResult.second)
            currentResult?.let {
                it.dataOrigin = DataOrigin.INTERNET
                getInstance().notifyResult(LyricsResultChange.Data(mediaInfo, it))
                return LyricUtils.parseLyric(it, mediaInfo, packageName)
            }
        }
        return null
    }

    private fun searchLyricsResultByInfo(mediaInfo: MediaInfo, originId: Long, sysLang: String) {
        var bestMatchSource: String? = null
        val bestMatchDistance: Long = 0
        for (provider in providers) {
            try {
                val lyricResult = provider.getLyric(mediaInfo)
                if (lyricResult?.lyric != null) {
                    val allLyrics = if (lyricResult.translatedLyric != null) {
                        LyricUtils.getAllLyrics(false, lyricResult.translatedLyric)
                    } else {
                        LyricUtils.getAllLyrics(false, lyricResult.lyric)
                    }
                    if (!CheckLanguageUtil.isJapanese(allLyrics)) {
                        when (sysLang) {
                            "zh-CN" -> if (lyricResult.translatedLyric != null) {
                                lyricResult.translatedLyric = ZhConverterUtil.toSimple(lyricResult.translatedLyric)
                            } else {
                                lyricResult.lyric = ZhConverterUtil.toSimple(lyricResult.lyric)
                            }

                            "zh-TW" -> if (lyricResult.translatedLyric != null) {
                                lyricResult.translatedLyric =
                                    ZhConverterUtil.toTraditional(lyricResult.translatedLyric)
                            } else {
                                lyricResult.lyric = ZhConverterUtil.toTraditional(lyricResult.lyric)
                            }

                            else -> {}
                        }
                    }
                    insertResData(originId, lyricResult)
                    if (LyricSearchUtil.isLyricContent(lyricResult.lyric)
                        && (bestMatchSource == null || bestMatchDistance > lyricResult.distance)
                    ) {
                        bestMatchSource = lyricResult.source
                    }
                }
            } catch (e: IOException) {
                e.fillInStackTrace()
            }
        }
        if (bestMatchSource != null) {
            val bestMatchResult = checkNotNull(getResByOriginIdAndProvider(originId, bestMatchSource))
            insertActiveLog(originId, bestMatchResult.id)
        }
    }
}
